### ISSUE TITLE
[qt5] Locating the Qt5 "Platform Plugin"

### ISSUE THREAD

**Issue by mitza-oci** [microsoft/vcpkg#7461]
A Qt application built with vcpkg-installed Qt libraries fails at startup because the "Platform Plugin" can't be located.

This was reported as a comment to a closed issue months ago (https://github.com/Microsoft/vcpkg/issues/2857#issuecomment-491060871) but it seems to be its own issue.

**Is your feature request related to a problem? Please describe.**

1. Installed Qt5 using vcpkg (qt5-base 5.12.3-3 along with other qt5-* pkgs needed for step 2: qt5-declarative qt5-modularscripts qt5-multimedia qt5-svg qt5-tools qt5-winextras
2. Configured a build of wireshark (3.0.3) to use the Qt5 from step 1; compiled it
3. Running `wireshark.exe` asserts with `This application failed to start because no Qt platform plugin could be initialized.`

**Proposed solution**
Step 3 should result in a running Wireshark process.  How best to do this will need to be determined by the vcpkg and Qt experts.

**Describe alternatives you've considered**
Qt plugins can be redirected using qt.conf files, but this doesn't seem to be the case for the *platform plugin*.  The *platform plugin* is loaded from the directory named `platforms` relative to the application executable.  Contrary to some references, it doesn't use `plugins\platforms` (for Qt 5.12).  This is distinct to *general Qt plugin* loading which is more flexible.  To override the *platform plugin* location, the environment variable `QT_QPA_PLATFORM_PLUGIN_PATH` can be set to any location.  As a workaround for vcpkg, that can be set to `%VCPKG%\installed\x64-windows\debug\plugins\platforms` (or the equivalent without `debug`).

**Additional context**
http://www.tripleboot.org/?p=536
https://doc.qt.io/qt-5/qt-conf.html
https://doc.qt.io/qt-5/windows-deployment.html#qt-plugins

**Comment by heydojo** [comment]
https://doc.qt.io/qt-5/windows-deployment.html

You should be using windeployqt if your platform target is Windows but unfortunately windeployqt does not work as expected when using Qt compiled via vcpkg.
windeployqt is typically the tool which will bundle all of the parts of Qt you need into one place.

Try:
- Compiling your application against Qt opensource pre-compiled download from Qt.
- Use windeployqt once the application has been compiled, to copy all the applications deps into a location containing a copy of the compiled application.
- Note down exactly which files have been copied and their exact locations.
- Use vcpkg to compile all of the parts of Qt that your application needs. And then compile your application against those same vcpkg Qt compiled libs (instead of Qt opensource.)
- Using a script and the information gained from the third step, copy the Qt files your application needs from your vcpkg tree and bundle them in with the application using a COPY command (or XCOPY.)
- Run windeployqt on the folder containing all of the copied files, something like so:

``` poweshell
SET PATH=%PATH%;%QTDIR%\bin;%QTDIR%\tools\qt5
%VCPKGDIR%\installed\x64-windows\tools\qt5-tools\windeployqt ^
-printsupport ^
-sql ^
--libdir %INSTDIR% ^
--plugindir %INSTDIR% ^
--release ^
--no-opengl-sw ^
--no-translations ^
--no-compiler-runtime ^
%INSTDIR%
```

YMMV.

**Comment by jsutes** [comment]
It's been awhile, but this is still an issue for me (I'm the one who made the referenced comment https://github.com/microsoft/vcpkg/issues/2857#issuecomment-491060871).

Here is the commit when the issue started: https://github.com/microsoft/vcpkg/commit/a1fe9544fd042a40fcfd28f04032a8f539486bdf

In qtdeploy.ps1, there are 5 lines of code (lines 18, 20, 21, 35, and 37) that have a "plugins" directory.  When I remove the "plugins" level from each path, then my qt deployments work as expected.  Otherwise, I get the "no Qt platform plugin could be initialized" error when I try to start my deployed application.

I don't know why the plugins directory level was added.

**Comment by JackBoosY** [comment]
@Neumann-A Could you please take a look?

Thanks.

**Comment by Neumann-A** [comment]
I think the idea was to also deploy a `qt.conf` with the corrected paths but I never touched `qtdeploy.ps1`. It probably requires an update or should be replaced `<platform>deployqt` if available. Theoretically we don't need to deploy qt dll we just need a qt.conf with the correct paths to make it work. (I mean the reason why VCPKG copies DLLs is probably just to make it work in VS via the run command)

**Comment by jsutes** [comment]
Thank you @JackBoosY  and @Neumann-A.
For my workflow, the DLLs get copied through the cmake install target.  If I'm doing a run command from within Visual Studio and my build directory, the DLLs do not appear to be copied yet. VS is able to find them without needing them to be copied to the build directory.
It's not until the cmake "install" command  is used that the DLLs are copied using the instructions in qtdeploy.ps1

The install/deploy step is important for us so we can distribute the final package to users with all the dependencies included.

**Comment by mitza-oci** [comment]
> 
> 
> I think the idea was to also deploy a `qt.conf` with the corrected paths but I never touched `qtdeploy.ps1`. It probably requires an update or should be replaced `<platform>deployqt` if available. Theoretically we don't need to deploy qt dll we just need a qt.conf with the correct paths to make it work. (I mean the reason why VCPKG copies DLLs is probably just to make it work in VS via the run command)

Please see the notes above under "Describe alternatives you've considered" regarding `qt.conf`.  Perhaps some of that has changed since I wrote it, but that's what I found out when creating this Issue.

**Comment by Kicer86** [comment]
I have the same issue.
During build I generate a tool which is use in later build steps.
It fails to work due to improper plugins path.

```qDebug() << QLibraryInfo::location(QLibraryInfo::PluginsPath);```
returns: 
```"C:/vcpkg/installed/x64-windows/tools/qt5/plugins"```
while it should be:
```C:\vcpkg\installed\x64-windows\tools\qt5\bin\plugins```

**Comment by JonLiu1993** [comment]
@mitza-oci ,Thanks for posting this issue. does this issue fixed? I can't reproduce this bug for now

**Comment by mitza-oci** [comment]
Yes, I think it was resolved somehow -- not sure if that was a change here or upstream.

I can now build and run Wireshark with a vcpkg-installed qt5 without doing anything special for the "Platform Plugin".  

My only remaining local patch is the following, and I don't think it's related to this issue:
```
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -2560,7 +2560,7 @@ if(BUILD_wireshark AND QT_FOUND)
 			POST_BUILD
 			COMMAND set "PATH=${QT_BIN_PATH};%PATH%"
 			COMMAND "${QT_WINDEPLOYQT_EXECUTABLE}"
-				--no-compiler-runtime
+				--no-compiler-runtime --no-angle
 				--verbose 0
 				$<$<BOOL:${MSVC}>:--pdb>
 				"$<TARGET_FILE:wireshark>"
```
